aiida/backends/tests/test_nodes.py

```diff
@@ -1702,100 +1621,385 @@ def test_deletion_non_existing_pk(self):
-        # I don't follow calls, so the slaves and their output are unharmed, as well as input
-        uuids_check_existence = [n.uuid for n in (in1, in2, slave1, outp1, slave2)]
-        # the wf and it's direct output and what it returned
-        uuids_check_deleted = [n.uuid for n in (wf, outp3, outp2, outp4)]
 
+        # By default deleting and input of the selection process will delete
+        # the workflow that involves the selection procedure but not any
+        # process that uses non-deleted inputs...
+        di1, di2, di3, di4, do1, pws, pcs, pwm = self._create_looped_graph()
+        uuids_check_existence = [n.uuid for n in (di1, di2, di3, do1, pcs)]
+        uuids_check_deleted = [n.uuid for n in (di4, pws, pwm)]
         with Capturing():
-            delete_nodes([wf.pk], verbosity=2, force=True, follow_calls=False, follow_returns=True)
+            delete_nodes([di4.pk], verbosity=2, force=True)
+        self._check_existence(uuids_check_existence, uuids_check_deleted)
 
+        # If the deleted input is the chosen one, then that's another story
+        di1, di2, di3, di4, do1, pws, pcs, pwm = self._create_looped_graph()
+        uuids_check_existence = [n.uuid for n in (di2, di3, di4)]
+        uuids_check_deleted = [n.uuid for n in (di1, do1, pws, pcs, pwm)]
+        with Capturing():
+            delete_nodes([di1.pk], verbosity=2, force=True)
         self._check_existence(uuids_check_existence, uuids_check_deleted)
+
+        # Deleting the workflow that chooses inputs should not touch the data
+        # provenance either.
+        di1, di2, di3, di4, do1, pws, pcs, pwm = self._create_looped_graph()
+        uuids_check_existence = [n.uuid for n in (di1, di2, di3, di4, do1, pcs)]
+        uuids_check_deleted = [n.uuid for n in (pws, pwm)]
+        with Capturing():
+            delete_nodes([pws.pk], verbosity=2, force=True)
         self._check_existence(uuids_check_existence, uuids_check_deleted)
 
-    def test_deletion_with_returns_n_loops(self):
-        """
-        Setting up a simple loop, to check that the following doesn't go bananas.
-        """
-        in1, in2, wf = [orm.Data().store(), orm.Data().store(), orm.WorkflowNode()]
-        wf.add_incoming(in1, link_type=LinkType.INPUT_WORK, link_label='link1')
-        wf.add_incoming(in2, link_type=LinkType.INPUT_WORK, link_label='link2')
-        wf.store()
-        in2.add_incoming(wf, link_type=LinkType.RETURN, link_label='link3')
+        # Again, deleting with call_calc/work_forward also deletes other parts of
+        # the graph.
+        di1, di2, di3, di4, do1, pws, pcs, pwm = self._create_looped_graph()
+        uuids_check_existence = [n.uuid for n in (di1, di2, di3, di4)]
+        uuids_check_deleted = [n.uuid for n in (do1, pws, pcs, pwm)]
+        with Capturing():
+            delete_nodes([pcs.pk], verbosity=2, force=True, call_calc_forward=True, call_work_forward=True)
```

Comment:
I think what you wanted to check is the following
```suggestion
            delete_nodes([pws.pk], verbosity=2, force=True, call_calc_forward=True, call_work_forward=True)
```
You can still keep both tests if you want, but the modified one is more important to test the flags, I would say.